annotations/tei_utils.py

```diff
@@ -0,0 +1,585 @@
+            # catch all namespace variations:
+            'xmlns="http://www.tei-c.org"', 
+            'xmlns="http://www.tei-c.org/ns/1.0"',
+            'xmlns:tei', 
+            'http://www.tei-c.org/ns/1.0'
+        ]):
+            is_tei = True
+
+    if is_tei:
+        content_type = 'text/xml+tei'
+    elif is_xml:
+        content_type = 'text/xml'
+    else:
+        content_type = 'text/plain'
+
+    return {
+        'is_xml': is_xml,
+        'is_tei': is_tei,
+        'content_type': content_type
+    }
+
+# ===== TEI Parsing Functions =====
+
+def parse_tei_document(xml_content):
+    """
+    Parse a TEI-XML document and prepare it for annotation.
+
+    Returns a dictionary with:
+        - original_xml
+        - clean_xml: namespace-stripped XML
+        - display_html: improved HTML representation
+        - element_map: mapping between display positions and TEI elements
+        - tei_metadata: extracted metadata from TEI header
+        - css: placeholder for CSS (if any)
+        - facsimile_data: info about facsimile images
+    """
+    try:
+        # Use a parser that removes processing instructions (XML declarations, etc.)
+        # This prevents issues with <?xml ...?> and <?xml-model ...?> tags 
+        parser = etree.XMLParser(remove_comments=True, remove_pis=True)
+        root = etree.fromstring(xml_content.encode('utf-8'), parser)
+
+        # Remove namespaces for easier processing
+        clean_root = strip_namespaces(root)
+        clean_xml = etree.tostring(clean_root, encoding='unicode')
+
+        # Extract metadata
+        metadata = extract_tei_metadata(clean_root)
+
+        # Extract facsimile data
+        facsimile_data = extract_facsimile_data(clean_root)
+
+        # Create display version
+        display_result = create_display_content(clean_root)
+
+        return {
+            'original_xml': xml_content,
+            'clean_xml': clean_xml,
+            'display_html': display_result['display_html'],
+            'element_map': display_result['element_map'],
+            'tei_metadata': metadata,
+            'css': display_result.get('css', ''),
+            'facsimile_data': facsimile_data
+        }
+    except Exception as e:
+        # Simple error reporting
+        raise ValueError(f"Failed to parse TEI-XML: {str(e)}")
+
+def strip_namespaces(node):
+    """
+    Deep-copy *node* with all namespaces removed.
+    • Ordinary element nodes are copied and cleaned.
+    • Processing instructions, comments, and text nodes that are
+      NOT children (.text / .tail) are skipped – they don't belong
+      in the cleaned tree and would otherwise raise TypeError
+      when appended.
+    """
+    # Skip non-element nodes (like processing instructions) to prevent errors
+    if not isinstance(node, etree._Element):
+        return None
+
+    # ── create namespace-free copy of *this* element ────────────────────
+    if node.tag.startswith('{'):
```

Comment:
when would a tag name start with `{`?

Reply:
ah I see, a better way is probably to do what the documentation suggests and to include namespaces when searching the xml: https://docs.python.org/3/library/xml.etree.elementtree.html#parsing-xml-with-namespaces